tuplex/codegen/src/BlockGeneratorVisitor.cc

```diff
@@ -5165,54 +5213,64 @@ namespace tuplex {
                 start = builder.CreateLoad(_env->CreateStructGEP(builder, exprAlloc.val, 0));
                 end = builder.CreateLoad(_env->CreateStructGEP(builder, exprAlloc.val, 1));
                 step = builder.CreateLoad(_env->CreateStructGEP(builder, exprAlloc.val, 2));
+            } else if(exprType.isIteratorType()) {
+                assert(forStmt->expression->hasAnnotation() && forStmt->expression->annotation().iteratorInfo);
+                iteratorInfo = forStmt->expression->annotation().iteratorInfo;
             } else {
                 fatal_error("unsupported expression type '" + exprType.desc() + "' in for loop encountered");
             }
             builder.CreateBr(condBB);
 
             // loop ending condition test
-            builder.SetInsertPoint(condBB);
-            llvm::PHINode *curr = builder.CreatePHI(_env->i64Type(), 2);
-            curr->addIncoming(start, entryBB);
-            curr->addIncoming(builder.CreateAdd(curr, step), iterEndBB);
+            llvm::PHINode *curr = nullptr;
             llvm::Value *loopCond = nullptr;
-            if(exprType == python::Type::RANGE) {
-                // step can be negative in range
-                loopCond = builder.CreateICmpSLT(builder.CreateMul(curr, step), builder.CreateMul(end, step));
+            builder.SetInsertPoint(condBB);
+            if(!exprType.isIteratorType()) {
```

Comment:
some comments for this code would be also appreciated. What's being checked here? what branches to generate?